test/test_sympy_utils.py

```diff
@@ -100,32 +100,47 @@ def generate_range(vals):
 
 class TestValueRanges(TestCase):
     @parametrize("fn", UNARY_OPS)
-    def test_unary_ref(self, fn):
+    @parametrize("dtype", ("int", "float"))
+    def test_unary_ref(self, fn, dtype):
+        dtype = {"int": sympy.Integer, "float": sympy.Float}[dtype]
         for v in CONSTANTS:
             if not valid_unary(fn, v):
                 continue
             with self.subTest(v=v):
-                ref_r = getattr(ReferenceAnalysis, fn)(sympy.Integer(v))
-                r = getattr(ValueRangeAnalysis, fn)(ValueRanges.wrap(v))
+                v = dtype(v)
+                ref_r = getattr(ReferenceAnalysis, fn)(v)
+                r = getattr(ValueRangeAnalysis, fn)(v)
+                self.assertEqual(r.lower.is_integer, r.upper.is_integer)
                 self.assertEqual(r.lower, r.upper)
+                self.assertEqual(ref_r.is_integer, r.upper.is_integer)
                 self.assertEqual(ref_r, r.lower)
 
     def test_pow_half(self):
         ValueRangeAnalysis.pow(ValueRanges.unknown(), ValueRanges.wrap(0.5))
 
     @parametrize("fn", BINARY_OPS)
-    def test_binary_ref(self, fn):
+    @parametrize("dtype_a", ("int", "float"))
+    @parametrize("dtype_b", ("int", "float"))
+    def test_binary_ref(self, fn, dtype_a, dtype_b):
+        to_dtype = {"int": sympy.Integer, "float": sympy.Float}
+        dtype_a = to_dtype[dtype_a]
+        dtype_b = to_dtype[dtype_b]
         for a, b in itertools.product(CONSTANTS, repeat=2):
             if not valid_binary(fn, a, b):
                 continue
+            a = dtype_a(a)
+            b = dtype_b(b)
             with self.subTest(a=a, b=b):
-                ref_r = getattr(ReferenceAnalysis, fn)(
-                    sympy.Integer(a), sympy.Integer(b)
-                )
-                r = getattr(ValueRangeAnalysis, fn)(
-                    ValueRanges.wrap(a),
-                    ValueRanges.wrap(b),
-                )
+                r = getattr(ValueRangeAnalysis, fn)(a, b)
+                if r == ValueRanges.unknown():
+                    continue
+                ref_r = getattr(ReferenceAnalysis, fn)(a, b)
+
+                # These min/max ops have a dtype dependent that's value-dependent on Python
```

Comment:
As I've mentioned to @nkaretnikov, the correct fix is to NOT use Python semantics, but instead have an alternate min/max implementation that always promotes based on a static promotion behavior.

Reply:
Done. Alas the current implementation doesn't work for `sympy.Expr`s, but this part of the code is not executed for dynamic shapes, so that's fine. I'll as Aaron how to fix this for Expressions.